api/src/main/java/jakarta/persistence/criteria/NumberExpression.java

```diff
@@ -0,0 +1,407 @@
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+/**
+ * Type for number query expressions.
+ *
+ * @param <X> the type of the expression
+ *
+ * @since 3.2
+ */
+@SuppressWarnings("hiding")
+public interface NumberExpression<X extends Number> extends Expression<X> {
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * greater than the first argument.
+     * @param y  expression
+     * @return greater-than predicate
+     * @see CriteriaBuilder#gt(Expression, Expression)
+     */
+    Predicate gt(Expression<? extends X> y);
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * greater than the first argument.
+     * @param y  value
+     * @return greater-than predicate
+     * @see CriteriaBuilder#gt(Expression, Number)
+     */
+    Predicate gt(X y);
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * greater than or equal to the first argument.
+     * @param y  expression
+     * @return greater-than-or-equal predicate
+     * @see CriteriaBuilder#ge(Expression, Expression)
+     */
+    Predicate ge(Expression<? extends X> y);
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * greater than or equal to the first argument.
+     * @param y  value
+     * @return greater-than-or-equal predicate
+     * @see CriteriaBuilder#ge(Expression, Number)
+     */
+    Predicate ge(X y);
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * less than the first argument.
+     * @param y  expression
+     * @return less-than predicate
+     * @see CriteriaBuilder#lt(Expression, Expression)
+     */
+    Predicate lt(Expression<? extends X> y);
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * less than the first argument.
+     * @param y  value
+     * @return less-than predicate
+     * @see CriteriaBuilder#lt(Expression, Number)
+     */
+    Predicate lt(X y);
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * less than or equal to the first argument.
+     * @param y  expression
+     * @return less-than-or-equal predicate
+     * @see CriteriaBuilder#le(Expression, Expression)
+     */
+    Predicate le(Expression<? extends X> y);
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * less than or equal to the first argument.
+     * @param y  value
+     * @return less-than-or-equal predicate
+     * @see CriteriaBuilder#le(Expression, Number)
+     */
+    Predicate le(X y);
+
+    /**
+     * Create a predicate for testing whether this expression is
+     * between the first and second arguments in value.
+     * @param x  expression
+     * @param y  expression
+     * @return between predicate
+     * @see CriteriaBuilder#between(Expression, Expression, Expression)
+     */
+    Predicate between(Expression<? extends X> x, Expression<? extends X> y);
```

Comment:
Is it really important the types are the same here? Why not allow comparison between floats and integers such as `5.0 between 4 and 6`? Or compare an integer against a long: `4 < Long.MAX_VALUE`?

This is why in Querydsl many of the methods on ComparableExpressions are typed like this:

```java
<A extends Number & Comparable<?>> BooleanExpression gt(A right)
```